File: Sources/LoginRadiusSDK.m
//
//  LoginRadiusSDK.m
//
//  Copyright © 2016 LoginRadius Inc. All rights reserved.
//

#import "LoginRadiusSDK.h"
#import "LoginRadiusSocialLoginManager.h"
#import "LRTouchIDAuth.h"
#import "LRFaceIDAuth.h"
#import "LRSession.h"
#import "LoginRadiusEncryptor.h"


static NSString * const LoginRadiusPlistFileName = @"LoginRadius";
static NSString * const LoginRadiusAPIKey = @"apiKey";
static NSString * const LoginRadiusRegistrationSource = @"registrationSource";
static NSString * const LoginRadiusCustomHeaders = @"customHeaders";
static NSString * const LoginRadiusSiteName = @"siteName";
static NSString * const LoginRadiusVerificationUrl = @"verificationUrl";
static NSString * const LoginRadiusKeychain = @"useKeychain";
static NSString * const LoginRadiusCustomDomain = @"customDomain";
static NSString * const LoginRadiusSetEncryption = @"setEncryption";



@interface LoginRadiusSDK ()
@property (strong, nonatomic) LoginRadiusSocialLoginManager *socialLoginManager;
@property (strong, nonatomic) LRTouchIDAuth *touchIDManager;
@property (strong, nonatomic) LRFaceIDAuth *faceIDManager;
@end

@implementation LoginRadiusSDK

- (instancetype)init {
    NSString *path = [[NSBundle mainBundle] pathForResource:LoginRadiusPlistFileName ofType:@"plist"];
    NSDictionary* values = [NSDictionary dictionaryWithContentsOfFile:path];
    NSString *apiKey = values[LoginRadiusAPIKey];
    NSString *registrationSource = values[LoginRadiusRegistrationSource];
    NSDictionary *customHeaders = values[ LoginRadiusCustomHeaders];
    NSString *siteName = values[LoginRadiusSiteName];
    NSString *verificationUrl = values[LoginRadiusVerificationUrl] ? values[LoginRadiusVerificationUrl] : @"https://auth.lrcontent.com/mobile/verification/index.html";
    BOOL useKeychain = values[LoginRadiusKeychain] ? [values[LoginRadiusKeychain] boolValue] : NO; // if nil set to false
    
    BOOL setEncryption = values[LoginRadiusSetEncryption] ? [values[LoginRadiusSetEncryption] boolValue] : NO; // if nil set to false
    
    
    NSString *customDomain = values[LoginRadiusCustomDomain] ? values[LoginRadiusCustomDomain] : @"";
    
    if([apiKey isEqualToString:@"<Your LoginRadius ApiKey>"] || [apiKey isEqualToString:@""]){
        NSString *str = nil;
        NSAssert(str, @"apiKey or siteName cannot be null in LoginRadius.plist");
    }
    
    NSAssert(apiKey, @"apiKey cannot be null in LoginRadius.plist");
    NSAssert(siteName, @"siteName cannot be null in LoginRadius.plist");
    
    //apiKey is stored in secure enclave if setEncryption set to true in LoginRadius.plist .
    if(setEncryption == true){
        
        NSData *key = [apiKey dataUsingEncoding:NSUTF8StringEncoding];
        NSData *decr = [[LoginRadiusEncryptor sharedInstance]EncryptDecryptText:key];
        
        NSString *myString;
        myString = [[NSString alloc] initWithData:decr encoding:NSASCIIStringEncoding];
    }
    
    if(!registrationSource){
        registrationSource = @"iOS";
    }
    
    self = [super init];
    
    if (self) {
        _apiKey = apiKey;
        _registrationSource = registrationSource;
        _siteName = siteName;
        _verificationUrl=verificationUrl;
        _customDomain =customDomain;
        _useKeychain = useKeychain;
        _session = [[LRSession alloc] init];
        _socialLoginManager = [[LoginRadiusSocialLoginManager alloc] init];
        _touchIDManager = [[LRTouchIDAuth alloc] init];
        _faceIDManager = [[LRFaceIDAuth alloc] init];
        _customHeaders = customHeaders;
        _setEncryption = setEncryption;
    }
    
    return self;
}

+ (instancetype)sharedInstance {
    static dispatch_once_t onceToken;
    static LoginRadiusSDK *instance;
    dispatch_once(&onceToken, ^{
        instance = [LoginRadiusSDK instance];
    });
    
    return instance;
}

+ (instancetype)instance {
    
    return [[LoginRadiusSDK alloc] init];
}


+ (BOOL) logout {
    
    BOOL is = [(LRSession *)[[self sharedInstance] session] logout];
    [[LoginRadiusSocialLoginManager sharedInstance] logout];
    // Clearing all stored tokens userprofiles for loginradius
    NSHTTPCookieStorage *storage = [NSHTTPCookieStorage sharedHTTPCookieStorage];
    for (NSHTTPCookie *cookie in [storage cookies]) {
        [storage deleteCookie:cookie];
    }
    return is;
}

+ (NSString*) apiKey {
    return [LoginRadiusSDK sharedInstance].apiKey;
}
+ (NSString*) registrationSource {
    return [LoginRadiusSDK sharedInstance].registrationSource;
}
+ (NSString*) siteName {
    return [LoginRadiusSDK sharedInstance].siteName;
}

+ (NSString*) verificationUrl {
    return [LoginRadiusSDK sharedInstance].verificationUrl;
}

+ (NSString*) customDomain {
    return [LoginRadiusSDK sharedInstance].customDomain;
}

+ (BOOL) useKeychain {
    return [LoginRadiusSDK sharedInstance].useKeychain;
}



+ (NSDictionary*) customHeaders {
    return [LoginRadiusSDK sharedInstance].customHeaders;
}

+ (BOOL) setEncryption {
    return [LoginRadiusSDK sharedInstance].setEncryption;
}

#pragma mark Application Delegate methods

- (void)applicationLaunchedWithOptions:(NSDictionary *)launchOptions {
    [self.socialLoginManager applicationLaunchedWithOptions:launchOptions];
    
    
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    
    return [[LoginRadiusSocialLoginManager sharedInstance] application:application openURL:url sourceApplication:sourceApplication annotation:annotation];
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
}


@end

File: Sources/LoginRadiusSDK.swift
//
//  LoginRadiusSDK.swift
//  LoginRadiusSwiftSDK
//
//  Created by Megha Agarwal.
//  Copyright © 2023 LoginRadius Inc. All rights reserved.
//




import Foundation
import UIKit

public typealias LRAPIResponseHandler = (_ data: [String: Any]?, _ error: Error?) -> Void
public typealias LRServiceCompletionHandler = (_ success: Bool, _ error: Error?) -> Void

public let LoginRadiusPlistFileName = "LoginRadius"
public let LoginRadiusAPIKey = "apiKey"
public let LoginRadiusRegistrationSource = "registrationSource"
public let LoginRadiusCustomHeaders = "customHeaders"
public let LoginRadiusSiteName = "siteName"
public let LoginRadiusVerificationUrl = "verificationUrl"
public let LoginRadiusKeychain = "useKeychain"
public let LoginRadiusCustomDomain = "customDomain"
public let LoginRadiusSetEncryption = "setEncryption"

public class LoginRadiusSDK {
    public static let sharedInstance = LoginRadiusSDK()
    public var apiKey: String
    public var registrationSource: String
    public var customHeaders: [String: Any]
    public var siteName: String
    public var verificationUrl: String
    public var useKeychain: Bool
    public var customDomain: String
    public var setEncryption: Bool
    public var socialLoginManager: LoginRadiusSocialLoginManager
    public var biometricManager: LRBiometrics
    public var session: LRSession
    
    public init() {
        let path = Bundle.main.path(forResource: LoginRadiusPlistFileName, ofType: "plist")
        let values = NSDictionary(contentsOfFile: path!) as! [String: Any]
        apiKey = values[LoginRadiusAPIKey] as! String
        registrationSource = values[LoginRadiusRegistrationSource] as? String ?? "iOS"
        
        customHeaders = values[LoginRadiusCustomHeaders] as? [String: Any] ?? [:]
        
        siteName = values[LoginRadiusSiteName] as! String
        verificationUrl = values[LoginRadiusVerificationUrl] as? String ?? "https://auth.lrcontent.com/mobile/verification/index.html"
        useKeychain = values[LoginRadiusKeychain] as? Bool ?? false
        customDomain = values[LoginRadiusCustomDomain] as? String ?? ""
        setEncryption = values[LoginRadiusSetEncryption] as? Bool ?? false
        
        assert(apiKey != "<Your LoginRadius ApiKey>" && !apiKey.isEmpty, "apiKey or siteName cannot be null in LoginRadius.plist")
        assert(siteName != "", "siteName cannot be null in LoginRadius.plist")
        assert(apiKey != "", "apiKey cannot be null in LoginRadius.plist")
        
        
        session = LRSession()
        socialLoginManager = LoginRadiusSocialLoginManager()
        biometricManager = LRBiometrics()
        customHeaders = values[LoginRadiusCustomHeaders] as? [String: Any] ?? [:]
        
        
        if setEncryption {
            let key = apiKey.data(using: .utf8)!
            let decr = LoginRadiusEncryptor.sharedInstance.encryptDecryptText(data: key)
            let myString = String(data: decr, encoding: .ascii)
        }
    }
    
    public static func logout() -> Bool {
        let isLoggedOut = (self.sharedInstance.session as! LRSession).logout()
        LoginRadiusSocialLoginManager.sharedInstance.logout()
        
        let storage = HTTPCookieStorage.shared
        for cookie in storage.cookies ?? [] {
            storage.deleteCookie(cookie)
        }
        
        return isLoggedOut
    }
    
    public static func apiKey() -> String {
        return LoginRadiusSDK.sharedInstance.apiKey
        
    }
    
    public static func registrationSource() -> String {
        return LoginRadiusSDK.sharedInstance.registrationSource
    }
    
    public static func siteName() -> String {
        return LoginRadiusSDK.sharedInstance.siteName
    }
    
    public static func verificationUrl() -> String {
        return LoginRadiusSDK.sharedInstance.verificationUrl
        
    }
    
    public static func customDomain() -> String {
        return LoginRadiusSDK.sharedInstance.customDomain
    }
    
    public static func useKeychain() -> Bool {
        return LoginRadiusSDK.sharedInstance.useKeychain
    }
    
    public static func customHeaders() -> [String: Any] {
        return LoginRadiusSDK.sharedInstance.customHeaders
    }
    
    public static func setEncryption() -> Bool {
        return LoginRadiusSDK.sharedInstance.setEncryption
    }
    
    // MARK: Application Delegate methods
    
    public func applicationLaunchedWithOptions(_ launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        LoginRadiusSocialLoginManager.sharedInstance.applicationLaunched(withOptions: launchOptions)
    }
    
    public func application(_ application: UIApplication, open url: URL, sourceApplication: String?, annotation: Any) -> Bool {
        return LoginRadiusSocialLoginManager.sharedInstance.application(application, open: url, sourceApplication: sourceApplication, annotation: annotation)
    }
    
    public func applicationDidBecomeActive(_ application: UIApplication) {
    }
}
